service/controller/resource/azuremachinepoolconditions/conditionsubnetready.go

```diff
@@ -0,0 +1,111 @@
+	// Now let's first check ARM deployment state
+	subnetDeploymentName := getSubnetDeploymentName(azureMachinePool.Name)
+	isSubnetDeploymentSuccessful, err := r.checkIfDeploymentIsSuccessful(ctx, deploymentsClient, azureMachinePool, subnetDeploymentName, azureconditions.SubnetReadyCondition)
+	if err != nil {
+		return microerror.Mask(err)
+	} else if !isSubnetDeploymentSuccessful {
+		// in the deployment is not yet successful, the check method will set
+		// appropriate condition value.
+		return nil
+	}
+
+	// Deployment is successful, now let's check the actual resource.
+	subnetsClient, err := r.azureClientsFactory.GetSubnetsClient(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	azureCluster, err := helpers.GetAzureClusterFromMetadata(ctx, r.ctrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	subnetName := azureMachinePool.Name
+	subnet, err := subnetsClient.Get(ctx, azureCluster.Name, azureCluster.Spec.NetworkSpec.Vnet.Name, subnetName, "")
+	if IsNotFound(err) {
+		r.setSubnetNotFound(ctx, azureMachinePool, subnetName, azureconditions.SubnetReadyCondition)
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Note: Here we check if the subnet exists and that its provisioning state
+	// is succeeded. It would be good to also check network security group,
+	// routing table and service endpoints.
+	if subnet.ProvisioningState == provisioningStateSucceeded {
+		capiconditions.MarkTrue(azureMachinePool, azureconditions.SubnetReadyCondition)
+	} else {
+		r.setSubnetProvisioningStateNotSuccessful(ctx, azureMachinePool, subnetName, subnet.ProvisioningState, azureconditions.SubnetReadyCondition)
+	}
+
+	r.logConditionStatus(ctx, azureMachinePool, azureconditions.SubnetReadyCondition)
+	r.logDebug(ctx, "ensured condition %s", azureconditions.SubnetReadyCondition)
+	return nil
+}
+
+func getSubnetDeploymentName(subnetName string) string {
+	return fmt.Sprintf("%s-%s", subnetDeploymentPrefix, subnetName)
+}
+
+func (r *Resource) setSubnetNotFound(ctx context.Context, cr capiconditions.Setter, subnetName string, condition capi.ConditionType) {
+	message := "Subnet %s is not found"
+	messageArgs := subnetName
+	capiconditions.MarkFalse(
+		cr,
+		condition,
+		SubnetNotFoundReason,
+		capi.ConditionSeverityError,
+		message,
+		messageArgs)
+
+	r.logWarning(ctx, message, messageArgs)
+}
+
+func (r *Resource) setSubnetProvisioningStateNotSuccessful(ctx context.Context, cr capiconditions.Setter, subnetName string, provisioningState network.ProvisioningState, condition capi.ConditionType) {
+	message := "Subnet %s provisioning state is %s"
+	messageArgs := []interface{}{subnetName, provisioningState}
+	reason := SubnetProvisioningStatePrefix + string(provisioningState)
+
+	capiconditions.MarkFalse(
+		cr,
+		condition,
+		reason,
+		capi.ConditionSeverityWarning,
+		message,
+		messageArgs...)
+
+	r.logWarning(ctx, message, messageArgs...)
```

Comment:
Is this case also covering when the subnet is still being created? Not sure if we should log `warning` in a such expected scenario. WDYT?

Reply:
> Is this case also covering when the subnet is still being created?

That would be caught earlier in function, in the ARM deployment check when calling `checkIfDeploymentIsSuccessful`, so it would be unexpected that after that we find a subnet with provisioning state different than `Succeeded` after already checking the ARM deployment.

Reply:
I see, cool.